### PR TITLE
[FIO internal] arm64: zynqmp: add bootslot command

### PR DESCRIPTION
**Pull request by igoropaniuk**
Introduce bootslot cmd for updating persistent register, whose values
are used by imgsel tool.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

Please do not submit a Pull Request via github.  Our project makes use of
mailing lists for patch submission and review.  For more details please
see https://www.denx.de/wiki/U-Boot/Patches

### CODE REVIEW

**Review by Tim-Anderson — APPROVED** [review]
LGTM

File: arch/arm/mach-zynqmp/cmd_bootslot.c [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+#include <asm/io.h>
+#include <spi.h>
+#include <spi_flash.h>
+
+#define SYS_CHECKSUM_OFFSET		(0x3U)
+#define PERSISTENT_REG_SIZE		(0x20000U)
+#define ENV_VAR_PREFIX			"zynqmp."
+
+/* The below enums denote persistent registers in Qspi Flash */
+struct sys_persistent_state {
```
Comment (MrCry0) [inline arch/arm/mach-zynqmp/cmd_bootslot.c:19]: all members of this struct are limited to {0,1}. Can we use a packed struct or even just a regular uint var to store all these variables as storage of flags?

Comment (igoropaniuk) [reply arch/arm/mach-zynqmp/cmd_bootslot.c:19]: We can't change this struct,as `imgsel` app expects this format of data to be flashed on QSPI (this is the first bootloader preflashed to 0x0 offset on QSPI on Kria K260 Starter Kit) https://github.com/Xilinx/embeddedsw/tree/master/lib/sw_apps/imgsel

I've just re-implemented in U-Boot what is done for Linux userspace https://github.com/Xilinx/linux-image_update/blob/master/image_update.c

Comment (MrCry0) [reply arch/arm/mach-zynqmp/cmd_bootslot.c:19]: Ah, sorry, I missed your description :)

**Review by MrCry0 — APPROVED** [review]
LGTM

File: arch/arm/mach-zynqmp/cmd_bootslot.c [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	return ret;
+}
+
+static void print_persistent_info(void)
+{
+	printf("Image A: ");
+	if (boot_img_info.persistent_state.img_a_bootable == 0U) {
+		printf("Non Bootable\n");
+		env_set(ENV_VAR_PREFIX "image_a_bootable", "0");
+
```
Comment (ricardosalveti) [inline arch/arm/mach-zynqmp/cmd_bootslot.c]: Drop extra line.

File: arch/arm/mach-zynqmp/Kconfig [hunk 3]
```diff
@@ -167,6 +167,20 @@ config ZYNQMP_PSU_INIT_ENABLED
...
+	help
+	  Command for switching A/B boot slots.
+
+config ZYNQMP_IMGSEL_PERSISTENT_REGISTER
+	default 0x100000
+	hex
+	help
+	  This option provides an address of persistent register offset,
+	  which is used by imgsel tool for detecting current bootable
+	  slot (A/B).
```
Comment (ricardosalveti) [inline arch/arm/mach-zynqmp/Kconfig]: Would be nice to extend this a bit more to point to imgsel as that is where this gets configured.

File: arch/arm/mach-zynqmp/cmd_bootslot.c [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+	boot_img_info.persistent_state.img_a_bootable = slot_a;
+	boot_img_info.persistent_state.img_b_bootable = slot_b;
+	boot_img_info.persistent_state.requested_boot_img = requested_slot;
+	boot_img_info.checksum = calculate_checksum();
+
+	ret = spi_flash_erase(flash, CONFIG_ZYNQMP_IMGSEL_PERSISTENT_REGISTER,
+			      flash->sector_size);
+	if (ret)
+		printf("Erase operation to SPI flash failed\n");
```
Comment (ricardosalveti) [inline arch/arm/mach-zynqmp/cmd_bootslot.c:195]: Should we try a write in case of a failure during erase?

**Comment by igoropaniuk** [comment]
@ricardosalveti all comments addressed

**Review by ricardosalveti — APPROVED** [review]
LGTM

**Pull request merged**